### ISSUE TITLE
Aria-errormessages property: what void does it fill? Cannot require aria-invalid

### ISSUE THREAD

**Issue by spanchang** [w3c/aria#128]
Reference : The aria-errormessage attribute references another element that contains custom error message text. Authors MUST use aria-invalid in conjunction with aria-errormessage. 
<span id="msgID" aria-live="assertive" style="visibility:visible"> Invalid time:  the time must be between 9:00 AM and 5:00 PM" </span>

Comment:
Is there really a necessity for the aria-errormessages property? 
What gap does it fill by being introduced into the specs? Is this documented anywhere?
It is not just a special case of aria-describedby?
Additionally, requiring use of aria-invalid=true I think is a complete no-no IMO.
The error text (span tag)  in the above example in the specs itself begins with "Invalid time".  
Why then should a screen reader  user be bombarded with the word "invalid" a second time because of aria-invalid=true?
Even when the error message does not begin with "Invalid" or "Error:", the message associated with the form control says it all.
See example #1 in ARIA21 for a perfect  use case of aria-invalid. In this case no error message is really needed as the label text, required property and aria-invalid convey what is wrong.
In other cases, when a descriptive error message like in the span text above  identifies the field, the problem and sometimes even the fix, what valud do the words "invalid entry" or "invalid data" bring to the user experience?
I strongly  suggest that the aria-errormessages property be  withdrawn completely.
Thanks and best regards,
Sailesh Panchang

**Comment by MichielBijl** [comment]
More (background) information about aria-errormessage on the list: http://lists.w3.org/Archives/Public/public-pfwg/2015Jul/0027.html

**Comment by richschwer** [comment]
The reason for the error message separation is that we needed a vehicle to associate error messages with form elements in a web page. We have use cases with multiple form elements - each entry may be invalid with a separate error message. We also need to disambiguate the error message from the actual aria-describedby description which may provide information as to how to enter data into the form field.

**Comment by MichielBijl** [comment]
> Why then should a screen reader user be bombarded with the word "invalid" a second time because of aria-invalid=true?

To further elaborate on this, on November 25th I asked on the list [“Does double announced content bother screen reader users?”](http://lists.w3.org/Archives/Public/public-aria/2015Nov/0017.html). From the replies that followed, we can safely assume that a short message—or word in this—that gets duplicated is not an issue.

**Comment by richschwer** [comment]
AT vendors can avoid announcing the descriptions and the error messages at the same time. Double spoken content will definitely annoy screen reader users. :-)

**Comment by MichielBijl** [comment]
I meant the duplication of the word “invalid” only; both the description and the error message would be annoying for sure.

**Comment by spanchang** [comment]
>>The reason for the error message separation is that we needed a vehicle to associate >>error messages with form elements in a web page.
>>We have use cases with multiple form elements - each entry may be invalid with a >>separate error message.
>>We also need to disambiguate the error message from the actual aria-describedby >>description which may provide information as to how to enter data into the form field.

These requirements are not new. And this is not really an
accessibility problem. It is a UI design and usability issue. I am not
talking about making error text PD becausethat is already possible).
Often oinstructional text is provided for  a small percentage of the
controls on a form. And these may be  given by way of a tooltip when
control gets focus or through a separate help link or just displayed
next to the label or control.

If instructional text is brief, it can be part of the accessible name
or, else   a part of the accessible description.
The error message (if one is needed)  can be made the first
id-reference in the aria-describedby. And the instructional text if it
is really important  can be the next reference or  can be de-linked if
it is still displayed when errors are displayed.
And it is surely bad accessibility UI to double speak content. Will
the standard visual UI be acceptable if pieces of text are repeated
arbitrarily just for the heck of it?

AT makers have built support for aria-invalid. So any changes to the
ARIA specs should take that as a given. Are there serious flaws in the
manner AT has implemented or do you expect making a minor tweak will
be considered as a top priority issue by AT makers?
And tomorrow if some developers continue using aria-describedby and
provide  access to error messages and instructional text, will that be
an accessibility failure? Which SC?
This property is simply muddying the waters and developers will be
confused: when do I use this property and when do I use the other? And
error messaging can be implemented correctly or wrongly using either.
And UA makers will needlessly be required to spin their wheels to
support another property that does exactly the same thing as
aria-describedby.
I think, our energies will be better spent in urging UA makers  to
improve support for the good old native HTML title attribute for
keyboard users.
 Here are a few examples for error messaging. Certainly there can be many more.
http://mars.dequecloud.com/demo/form-alert2.htm
http://mars.dequecloud.com/demo/form-alert3.htm
http://mars.dequecloud.com/demo/Form-alertWithJQ.htm
http://mars.dequecloud.com/demo/Form-AriaLiveWithJQ.htm
Best regards,
Sailesh Panchang
Principal Accessibility Consultant, Deque Systems
Desk 703-225-0380 ext 105
Mobile 571-344-1765

**Comment by MichielBijl** [comment]
> The error message—if one is needed—can be made the first id-reference in the `aria-describedby`.

How do we indicate that this newly linked ID is now an error message?

**Comment by spanchang** [comment]
Are you saying today all Web content that is out there is failing SC
3.3.1? Are you saying that there is no mechanism  to convey  that a
field has failed validation, explain the problem and fix and do it in
a PD way today?
The text of the error message  conveys the issue else it is a
usability problem not an accessibility problem. And  it is a good
practice to prefix the message with "Error:" in some cases. And if an
error icon is being used like in one of the examples given, what more
is needed?
Also, I forgot to say in my last email: aria-errormessages  suggests
use of aria-live. Are you really suggesting this should be attached to
all 3 fields on a form that fail  validation? Or only to the global
alert message? Yes it can be done if only one (and not multiple) error
is displayed at a time.
It will really hellp if to get counter responses to all my arguments
instead of  just a piecemeal response. That will be a comprehensive
approach then.
Thanks,
Sailesh

**Comment by MichielBijl** [comment]
@spanchang Can you tell me what PD stands for? I understand it would be more helpful if I replied to all your arguments, but I can only reply to things I understand myself. To clarify, I am neither for nor against `aria-errormessage`. I'm a scientist; evidence shows me the way, and to this point, more evidence points at `aria-errormessage` being useful. Ultimately, this should be tested with users of course.

Pro | Con
------------ | -------------
More semantic | More work for UA / AT
Can have both error and description | Might be too verbose
Easier to find for developers new to accessibility and ARIA |


No, I'm not saying existing techniques fail 3.3.1. But I do think there is place for a more semantic way of doing it.

I think [Brian LePore put it rather nicely](https://gitter.im/w3c/a11ySlackers?at=56742d0ff99ceb846ae69f20):

> I see why this wouldn't really be needed as there are elements already that can be made to do what is necessary, but I think error states are common enough that having something explicit that does it would be of value.

**Comment by spanchang** [comment]
PD is programmatically determinable ... as used in SC 1.3.1 etc. of
WCAG 2. Sorry for not not having used the expanded form.
In my view there is zero or  negative value for the attribute as it
may create more problems as my previous emails suggest.
And I am serious when I say, our time may be better spent  coaxing
user agent makers to improve support for the title attribute so it is
available to a wider group of PWD. It will simplify the process of
providing accessible description (or name in some instances).
Thanks and best wishes,
Sailesh Panchang

**Comment by MichielBijl** [comment]
More information on why we need `aria-errormessage`: [James Nurthen on a11ySlackers](https://gitter.im/w3c/a11ySlackers?at=5676cedfb5777fb85ba4d381).

**Comment by spanchang** [comment]
If some screen readers in some situations do not read aria-describedby , it is something that needs to be fixed in the UA or AT. Regardless of whether the attribute is conveying an error or something else, it is problematic if the aria-describedby text is not announced. 
I have listed several reasons in preceding emails why aria-errormessages is not a good idea.
I have not really seen a single convincing  reason why a separate attribute is needed for  notifying error messages. Again I will appreciate if  you offer comprehensive reasoning for  this proposal  counter arguments to my statements.
Thanks,
Sailesh

**Comment by MichielBijl** [comment]
Jamie Teh has written [a great article on aria-describedby](http://blog.jantrid.net/2015/12/woe-aria-aria-describedby-to-report-or.html) this week that is well worth a read. He states that `aria-describedby` should _only_ be read when the user asks for it, that would disqualify it for the purpose `aria-errormessage` fills. Here are _my_ answers to all your questions in the opening post—that I can answer:

> Is there really a necessity for the aria-errormessages property? 

It's probably not a 100% necessary, but there are definitely cases where it makes sense to have this.

> What gap does it fill by being introduced into the specs?

A semantic way of binding an error message to an input that can exist alongside a description; as pointed out in earlier comments[[1]](https://github.com/w3c/aria/issues/128#issuecomment-165552095)[[2]](https://github.com/w3c/aria/issues/128#issuecomment-165564049)[[3]](https://github.com/w3c/aria/issues/128#issuecomment-165837141)[[4]](https://github.com/w3c/aria/issues/128#issuecomment-166151527), but I'll try to sum up some here:

* More semantic way to define the relationship;
* Can have both error and description;
* `aria-describedby` should—according to some—not bespoken until the user indicates that they need it;
* Easier to find for developers new to accessibility and ARIA;
* We needed a vehicle to associate error messages with form elements in a web page.

> Is this documented anywhere?

Not sure if there is anything else, but here is the spec: [WAI-ARIA 1.1: `aria-errormessage` (property)](http://www.w3.org/TR/wai-aria-1.1/#aria-errormessage)

> It is not just a special case of aria-describedby?

Yes, in a way.

> Additionally, requiring use of aria-invalid=true I think is a complete no-no IMO. The error text (span tag) in the above example in the specs itself begins with "Invalid time".

I'm not sure this is really an issue, but we would have to ask assistive technology users. From what I've heard, this isn't issue. Besides. What happens if the error message doesn't start with “invalid” is it still a problem then? Might as well start with “Error:” or might not include such a word at all.

* * *

I hope that other people will join this conversation and share their views.

**Comment by StommePoes** [comment]
This reminds me of the difference between (modal) dialogs and alertdialogs. The difference is the content. But people want the alertdialogs to behave differently due to the content.

Wonder if we could get some sites like AccessibleApps (Q) to implement some of these things (temporarily) for testing and get real-world user feedback.

**Comment by MichielBijl** [comment]
@spanchang could you explain what the problem with `aria-errormessage` is? I'm not sure I understand. Is that we're duplicating what we can do with `aria-describedby` today? Or that this is more work for UA/AT vendors? Or maybe something completely different?

**Comment by MichielBijl** [comment]
We've been [discussing this for a bit on a11ySlackers](https://gitter.im/w3c/a11ySlackers?at=56783de1091b6f9e043a1294) and were looking for other ways to accomplish this and what the issue with those would be. I'm wondering whether there was ever a proposal for an `<error>` element, and if so, why it was rejected? It could function the same as a `<label>`. So for example:

```
<input id="phone-number" type="tel" invalid>
<error for="phone-numer">Invalid value: only digits are allowed.</error>
```

This could be build into ARIA as a role:

```
<p role="error">Error message</p>
```

This seems like a very logical thing to add to ARIA/HTML. It is linked in the same way, but has different semantics. Are there any issues with this approach?

**Comment by spanchang** [comment]
Yes it is duplication and it surely does not solve any accessibility
problem today because all forms that comply with SC 3.3.1 and 1.3.1
(A) do so without the need for any additional properties / attributes.
Take a look at  this page (created  exactly 2 years ago) and
referenced in an earlier email:
Enter just  2 or 3 digits or non-numeric data in the PIN field, enter
date as  9/5/12 and submit the form at
http://mars.dequecloud.com/demo/form-alert3.htm
Check how the errors are conveyed for missing data, invalid data etc.
Also experiment with a date in proper format which is in the past.
Please read through the  explanatory text on the page for notes on AT
support, use of only aria-invalid  or use of programmatically
associated (via aria-describedby) error text.
My previous emails describe the objections to aria-errormessage too:
- duplication
- problematic to use with aria-invalid because the error text alerts
user to the error and aria-invalid is redundant
Problematic to use with aria-live except when a single error is
displayed at a time and there is no global error message for the form
- It is not the goal of ARIA to help developers track error messages
with form controls they do it via other means and ARIA is not designed
for that
- You argue that aria-describedby is not spoken by screen readers
sometimes. Any guarantee that UA-AT combos will not create a similar
problem with a new attribute? Is it not better to troubleshoot the
problem for aria-describedby with AT and UA makers and get the issue
addressed if there is a flaw?

So when essential  instructional text related to a form control needs
to be conveyed, it can be done so using the label or an accessible
tooltip or placeholder attribute. One can even use two label elements
that point to the same form control using for-id but then
aria-describedby  has more robust support.
Programmatic association is needed when the association or
relationship is conveyed visually by presentation. Then one can use
aria-describedby or other aria method .
Else, the text does not have to be programmatically associated.
When label text conveys field name, data format with the asterisk for
required field, just aria-invalid is sufficient to convey that entered
data is invalid when not in the right format. Yet in this example
For missing data, the adjacent text is associated via aria-describedby
 to aid the user perhaps because he has not realized from the asterisk
that it is required.
Note: I have not deliberately used required / aria-required  here.

I guess I have covered all of what I said in the earlier emails
without sounding repetitive. I request you to review those again too.

Best wishes and regards,
Sailesh Panchang

**Comment by michael-n-cooper** [comment]
It's possible this has been discussed to death already, but not clear there would be consensus to close, therefore moving to ARIA 1.2 for later processing.

**Comment by michael-n-cooper** [comment]
Moving back to ARIA 1.1 and closing, Rich says this was dealt with in the WG.

**Comment by richschwer** [comment]
This was discussed and explained to death and we have two platform implementations under way.
 
It was also put out for concensus review.
 
Something that has to stop is people coming back to the group and rehashing old issues especially after so much time was spent on this. If you want to know why we have it in the spec. then go back to the meetings to see the discussions. You can follow them on the aria discussion as well as the accessibility api mapping discussion.
 
Close this issue.
 
Rich Schwerdtfeger
 
 
----- Original message -----From: Michiel Bijl <notifications@github.com>To: w3c/aria <aria@noreply.github.com>Cc: Richard Schwerdtfeger/Austin/IBM@IBMUSSubject: Re: [aria] Aria-errormessages property: what void does it fill? Cannot require aria-invalid (#128)Date: Mon, Dec 21, 2015 11:44 AM 
@spanchang could you explain what the problem with aria-errormessage is? I'm not sure I understand. Is that we're duplicating what we can do with aria-describedby today? Or that this is more work for UA/AT vendors? Or maybe something completely different?
